packages/optimizer/lib/transformers/MinifyHtml.js

```diff
@@ -0,0 +1,172 @@
+const COMMENT_DEFAULT_IGNORE = /^\s*__[a-bA-Z0-9_-]+__\s*$/;
+
+// Matches all consecutive whitesapce
+const WHITESPACE_REGEX = /\s+/g;
+
+/**
+ * MinifyHtml - minifies files size by:
+ *
+ * - minifying inline JSON
+ * - minifying inline amp-script using https://www.npmjs.com/package/terser
+ * - collapsing whitespace outside of pre, script, style and area.
+ * - removing comments
+ *
+ * This transformer supports the following options:
+ *
+ * - `minify [Boolean]`: Enables HTML minification. The default is `true`.
+ */
+class MinifyHtml {
+  constructor(config) {
+    this.opts = {
+      minify: !!config.minify || true,
+      minifyAmpScript: true,
+      minifyJSON: true,
+      collapseWhitespace: true,
+      removeComments: true,
+      canCollapseWhitespace: true,
+      canTrimWhitespace: true,
+      commentIgnorePattern: COMMENT_DEFAULT_IGNORE,
+    };
+    this.log = config.log.tag('MinifyHtml');
+  }
+  transform(tree) {
+    if (!this.opts.minify) {
+      return;
+    }
+    // store nodes for later deletion to avoid changing the tree structure
+    // while iterating the DOM
+    const nodesToRemove = [];
+    // recursively walk through all nodes and minify if possible
+    this.minifyNode(tree, this.opts, nodesToRemove);
+    for (const node of nodesToRemove) {
+      remove(node);
+    }
+  }
+
+  minifyNode(node, opts, nodesToRemove) {
+    if (node.type === 'text') {
+      this.minifyTextNode(node, opts, nodesToRemove);
+    } else if (node.type === 'comment') {
+      this.minifyCommentNode(node, opts, nodesToRemove);
+    } else if (node.tagName === 'script') {
+      this.minifyScriptNode(node, opts);
+    }
+    // update options based on the current node
+    const childOpts = Object.assign({}, opts);
+    if (opts.canCollapseWhitespace && !this.canCollapseWhitespace(node.tagName)) {
+      childOpts.canCollapseWhitespace = false;
+    }
+    if (opts.canTrimWhitespace && !this.canTrimWhitespace(node.tagName)) {
+      childOpts.canTrimWhitespace = false;
+    }
+    // minify all child nodes
+    for (const child of node.children || []) {
+      this.minifyNode(child, childOpts, nodesToRemove);
+    }
+  }
+
+  minifyTextNode(node, opts, nodesToRemove) {
+    if (!node.data || !opts.collapseWhitespace) {
+      return;
+    }
+    if (opts.canCollapseWhitespace) {
+      node.data = node.data.replace(WHITESPACE_REGEX, ' ');
+    }
+    if (opts.canTrimWhitespace) {
+      node.data = node.data.trim();
+    }
+    // remove empty nodes
+    if (node.data.length === 0) {
+      nodesToRemove.push(node);
+    }
+  }
+
+  minifyCommentNode(node, opts, nodesToRemove) {
+    if (!node.data || !opts.removeComments) {
+      return;
+    }
+    if (opts.commentIgnorePattern.test(node.data)) {
+      return;
+    }
+    nodesToRemove.push(node);
+  }
+
+  minifyScriptNode(node, opts) {
+    const isJson = this.isJson(node);
+    const isAmpScript = this.isInlineAmpScript(node);
```

Comment:
Shouldn't one of these exclude the other? No need to process the second if the first is already true.

Reply:
addressed in https://github.com/ampproject/amp-toolbox/pull/591/commits/62c343e5685670d67a5fd3e0ff348bf0867d3c89